File: src/chapters/linear-systems/content/code/tridiagonal.cpp
#include <vector>
#include <iostream>
using namespace std;
using Vec = vector<double>;

// Thomas algorithm for a tridiagonal system (a sub-, b diag, c super-, d rhs).
Vec thomas(const Vec& a, const Vec& b, Vec c, Vec d) {
    int n = d.size();
    c[0] /= b[0]; d[0] /= b[0];
    for (int i = 1; i < n; ++i) {
        double m = b[i] - a[i] * c[i - 1];
        if (i < n - 1) c[i] /= m;
        d[i] = (d[i] - a[i] * d[i - 1]) / m;
    }
    Vec x(n);
    x[n - 1] = d[n - 1];
    for (int i = n - 2; i >= 0; --i) x[i] = d[i] - c[i] * x[i + 1];
    return x;
}

int main() {
    Vec a = {0, -1, -1, -1}, b = {4, 4, 4, 4}, c = {-1, -1, -1, 0}, d = {2, 4, 6, 13};
    for (double v : thomas(a, b, c, d)) cout << v << " ";
    cout << "\n";
}

File: src/chapters/linear-systems/content/code/tridiagonal.rs
// Thomas algorithm for a tridiagonal system (a sub-, b diag, c super-, d rhs).
fn thomas(a: &[f64], b: &[f64], c: &[f64], d: &[f64]) -> Vec<f64> {
    let n = d.len();
    let (mut cc, mut dd) = (c.to_vec(), d.to_vec());
    cc[0] /= b[0]; dd[0] /= b[0];
    for i in 1..n {
        let m = b[i] - a[i] * cc[i - 1];
        if i < n - 1 { cc[i] /= m; }
        dd[i] = (dd[i] - a[i] * dd[i - 1]) / m;
    }
    let mut x = vec![0.0; n];
    x[n - 1] = dd[n - 1];
    for i in (0..n - 1).rev() { x[i] = dd[i] - cc[i] * x[i + 1]; }
    x
}
fn main() {
    let a = [0.0, -1.0, -1.0, -1.0];
    let b = [4.0, 4.0, 4.0, 4.0];
    let c = [-1.0, -1.0, -1.0, 0.0];
    let d = [2.0, 4.0, 6.0, 13.0];
    println!("{:?}", thomas(&a, &b, &c, &d));
}
